### PR TITLE
Port over cache doc

### LINKED ISSUE #59: Write documentation for the Cache component
We need to bring over the docs from the old docs relating to the Caching bundle, ensure that it's up to date, and write it in RST format.

https://developer.mautic.org/#cache-bundle

### PR DESCRIPTION
**Pull request by fakela**
Fixes #59

WIP

<a href="https://gitpod.io/#https://github.com/mautic/developer-documentation-new/pull/148"></a>

### CODE REVIEW

**Comment by RCheesley** [comment]
@fakela I think this is relevant for the current users (Mautic 4). We will need to do another PR for Mautic 5, because the whole system is changing.

When 5.0 General Availability is released we will branch from main -> 4.x and then the current 5.x branch will become main.

**Comment by fakela** [comment]
@RCheesley I have port over the cache docs that is in the dev documentation. If I am missing out on anything please let me know.


Also once this is confirmed as okay, I'll create another PR using the 5.x branch

**Comment by fakela** [comment]
I had a little issue previewing this so I stuck to using GitHub preview. I'll try to debug and see how to make it work as expected

**Comment by RCheesley** [comment]
Should be fixed with Gitpod now - was a problem with the application not being installed I think!

File: docs/components/cache.rst [hunk 1]
```diff
@@ -1,4 +1,122 @@
-Cache
+CACHE
 #####
 
+Symfony makes heavy use of a filesystem cache. Frequently clearing this cache will be required when developing for Mautic. By default, the cache is located in ``app/cache/ENV`` where ``ENV`` is the environment currently accessed (``dev`` or ``prod``). To rebuild the cache, the ``ENV`` can just be deleted or run the Symfony command ``php app/console cache:clear --env=ENV``. If a specific environment is not passed to the command via ``--env=ENV``, the ``dev`` environment will be used by default.
```
Comment (RCheesley) [inline docs/components/cache.rst:4]: Update for Mautic 3 and beyond (app/console -> bin/console, app/cache -> var/cache) and multiple Vale fixes.

```suggestion
Symfony makes heavy use of a filesystem cache. When developing for Mautic, clearing the cache is a regular occurrence. By default, Mautic instances have the cache located in ``var/cache/ENV`` where ``ENV`` is the environment currently accessed (``dev`` or ``prod``). To rebuild the cache, delete the relevant ``ENV`` folder within the cache directory, or run the Symfony command ``php bin/console cache:clear --env=ENV``. If a specific environment isn't passed to the command via ``--env=ENV``, Mautic uses the ``dev`` environment by default.
```

File: docs/components/cache.rst [hunk 2]
```diff
@@ -1,4 +1,122 @@
...
 
+In the ``dev`` environment, translations, views, and assets are not cached. However, changes to these files will require the cache to be cleared for them to take effect in the ``prod`` environment. Changes to Mautic config files, Symfony config files, etc., will require the cache to be cleared regardless of the environment.
+
+The typical rule of thumb is, if Mautic is not acting as you expect after making changes, try clearing your cache. If you get "class could not be found" or "cannot redeclare class" errors when using the ``cache:clear`` command, manually delete the ``app/cache/ENV`` folder then run the command and/or browse to the site to rebuild.
```
Comment (RCheesley) [inline docs/components/cache.rst:8]: Update the file path, and Vale fixes.

```suggestion
The typical rule of thumb is, if Mautic isn't acting as you expect after making changes, try clearing your cache. If you get ``class could not be found`` or ``cannot redeclare class`` errors when using the ``cache:clear`` command, manually delete the ``var/cache/ENV`` folder then run the command and/or browse to the site to rebuild.
```

File: docs/components/cache.rst [hunk 3]
```diff
@@ -1,4 +1,122 @@
...
+
+Cache Bundle
```
Comment (RCheesley) [inline docs/components/cache.rst:10]:
```suggestion
Cache bundle
```

File: docs/components/cache.rst [hunk 4]
```diff
@@ -1,4 +1,122 @@
...
+------------
+
+Enables PSR-6 and PSR-16 caching. Check: `Symfony Cache Component <https://symfony.com/doc/current/components/cache.html>`_
```
Comment (RCheesley) [inline docs/components/cache.rst:13]: @fakela please convert this to a link using the 'make link' command within the docs directory, and commit the file that is created. Thanks!

File: docs/components/cache.rst [hunk 5]
```diff
@@ -1,4 +1,122 @@
...
 
+Symfony makes heavy use of a filesystem cache. Frequently clearing this cache will be required when developing for Mautic. By default, the cache is located in ``app/cache/ENV`` where ``ENV`` is the environment currently accessed (``dev`` or ``prod``). To rebuild the cache, the ``ENV`` can just be deleted or run the Symfony command ``php app/console cache:clear --env=ENV``. If a specific environment is not passed to the command via ``--env=ENV``, the ``dev`` environment will be used by default.
 
+In the ``dev`` environment, translations, views, and assets are not cached. However, changes to these files will require the cache to be cleared for them to take effect in the ``prod`` environment. Changes to Mautic config files, Symfony config files, etc., will require the cache to be cleared regardless of the environment.
+
+The typical rule of thumb is, if Mautic is not acting as you expect after making changes, try clearing your cache. If you get "class could not be found" or "cannot redeclare class" errors when using the ``cache:clear`` command, manually delete the ``app/cache/ENV`` folder then run the command and/or browse to the site to rebuild.
+
+Cache Bundle
+------------
```
Comment (RCheesley) [inline docs/components/cache.rst:11]: Use the correct heading nesting.
```suggestion
************
```

File: docs/components/cache.rst [hunk 6]
```diff
@@ -1,4 +1,122 @@
...
+
+Enables PSR-6 and PSR-16 caching. Check: `Symfony Cache Component <https://symfony.com/doc/current/components/cache.html>`_
+
+Namespace versus tag
+---------------------
```
Comment (RCheesley) [inline docs/components/cache.rst:16]: Nesting the headings correctly

```suggestion
====================
```

File: docs/components/cache.rst [hunk 7]
```diff
@@ -1,4 +1,122 @@
...
+
+This bundle introduces tags to cache. All its adapters are fully tag aware which makes the use of namespace obsolete for daily use.
```
Comment (RCheesley) [inline docs/components/cache.rst:18]: I think this needs a 'the' here but maybe there is a reason why not to. Worth checking with the devs.
```suggestion
This bundle introduces tags to the cache. All its adapters are fully tag aware which makes the use of namespace obsolete for daily use.
```

File: docs/components/cache.rst [hunk 8]
```diff
@@ -1,4 +1,122 @@
...
+
+Previously, if you wanted to keep control on cache section and did not want to hold the index of all keys to clear, you would have to use namespace.
```
Comment (RCheesley) [inline docs/components/cache.rst:20]: Vale fix
```suggestion
Previously, if you wanted to keep control on cache section and didn't want to hold the index of all keys to clear, you would have to use namespace.
```

File: docs/components/cache.rst [hunk 9]
```diff
@@ -1,4 +1,122 @@
...
+
+Disadvantage of this approach is a new adapter being created for each namespace.
```
Comment (RCheesley) [inline docs/components/cache.rst:22]: Vale fix
```suggestion
The main disadvantage of this approach is that Mautic creates a new adapter for each namespace.
```

File: docs/components/cache.rst [hunk 10]
```diff
@@ -1,4 +1,122 @@
...
+
+Symfony 3.4 Cache uses tag-aware adapters. If you want to clear all records related to your bundle or component, you just need to tag them.
```
Comment (RCheesley) [inline docs/components/cache.rst:24]:
```suggestion
From Symfony 3.4, the cache uses tag-aware adapters. If you want to clear all records related to your Bundle or Component, you just need to tag them.
```

File: docs/components/cache.rst [hunk 11]
```diff
@@ -1,4 +1,122 @@
...
+    $item->tag(['firstTag', 'secondTag']);
+    $item->expiresAfter(20000);
+
+All you need to do now is to clear all tagged items:
+
+.. code-block:: php
+
+    $cache->invalidateTags(['firstTag']);
+
+POOLS CLEARING
```
Comment (RCheesley) [inline docs/components/cache.rst:43]:
```suggestion
Pools clearing
```

File: docs/components/cache.rst [hunk 12]
```diff
@@ -1,4 +1,122 @@
...
+==============
+
+Removing Cache Items
```
Comment (RCheesley) [inline docs/components/cache.rst:46]:
```suggestion
Removing cache items
```

File: docs/components/cache.rst [hunk 13]
```diff
@@ -1,4 +1,122 @@
...
+--------------------
+
+Cache Pools include methods to delete a cache item, some of them, or all of them. The most common is ``Psr\\Cache\\CacheItemPoolInterface::deleteItem``, which deletes the cache item identified by the given key.
+
+.. code-block:: php
+
+    $isDeleted = $cache->deleteItem('user_'.$userId);
+
+Use the ``Psr\\Cache\\CacheItemPoolInterface::deleteItems`` method to delete several cache items simultaneously (it returns true only if all the items have been deleted, even when any or some of them don’t exist):
```
Comment (RCheesley) [inline docs/components/cache.rst:55]:
```suggestion
Use the ``Psr\\Cache\\CacheItemPoolInterface::deleteItems`` method to delete several cache items simultaneously - it returns true only if all the items have been deleted, even when any or some of them don't exist:
```

File: docs/components/cache.rst [hunk 14]
```diff
@@ -1,4 +1,122 @@
...
+
+Configuration
+-------------
+
+Plugin comes preconfigured to utilize filesystem caching.
```
Comment (RCheesley) [inline docs/components/cache.rst:60]:
```suggestion
Plugins come preconfigured to utilize filesystem caching.
```

**Review by RCheesley — CHANGES_REQUESTED** [review]
Quite a lot of updating needed on this PR - it's outdated and there are some Vale fixes.

**Comment by RCheesley** [comment]
@fakela please review feedback here. Thanks!

**Comment by fakela** [comment]
> @fakela please review feedback here. Thanks!

done

**Comment by RCheesley** [comment]
This needs to be rebased against 5.x as it was based on main - can you do that @fakela ? Or it might be easier to start a new PR perhaps?

**Comment by fakela** [comment]
> This needs to be rebased against 5.x as it was based on main - can you do that @fakela ? Or it might be easier to start a new PR perhaps?

I'll create a new PR for it

**Pull request closed without merging**